### PR TITLE
WIP: Adding keyring 9.0

### PR DESCRIPTION
**Pull request by ericdill**

### CODE REVIEW

**Comment by jakirkham** [comment]
See this [comment]( https://github.com/conda-forge/staged-recipes/issues/666#issuecomment-221125814 ) for a workaround.

**Comment by ericdill** [comment]
@jaraco I am adding keyring v9.0 to conda-forge for linux, mac and windows. This will allow users to easily install this with [conda] (http://conda.pydata.org/docs/) (conda install -c conda-forge keyring). The maintenance burden here is light as all testing and releases are built and deployed automatically with CI. Changes to how it is being built can be controlled by changing the recipe which will be located at https://github.com/conda-forge/keyring-feedstock shortly after this PR is finished and merged. If you would like to be added as a maintainer (to make new releases for instance), I can happily add you to the maintainers list. If not, that is totally fine too!  Is there anyone else that you think would want to be pinged about maintaining this build?

**Comment by ericdill** [comment]
I am selfishly skipping windows builds as I need this package for linux only and am starting to go down a rabbit hole for trying to figure out how to get this building on windows.  The test suite for keyring seems to be preferentially executed with `python setup.py test` which.  On windows, this test execution requires `pywin32-ctypes` which has one release from 2014.  I was not successful at executing the test suite from that tag (v0.0.1) on python 3.5 which led me to the conclusion that this is a problem that I do not particularly care to solve since I only need this for linux (osx is a bonus).

So my question to the conda-forge community is basically this: Can I selfishly skip building conda packages for windows since I am running into some issues?

**Comment by jakirkham** [comment]
No complaints from me.

In fact, we regularly skipped Windows in the beginning because it proved to challenging for a package. We have become a bit better about trying, but not all of us have so much Windows experience. It seems like you made a decent effort so it is totally fine by me if you call it quits if it doesn't go easily.

One thought is you could drop a line in this issue ( https://github.com/conda-forge/staged-recipes/issues/161 ) to see if someone with more experience on Windows (and possibly a working Window dev environment) can help push it along. Though that can be post merge as far as I'm concerned.

File: recipes/keyring/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+package:
+    name: keyring
+    version: {{ version }}
+
+source:
+    fn: keyring-{{ version }}.tar.gz
+    url: https://files.pythonhosted.org/packages/source/k/keyring/keyring-{{ version }}.tar.gz
+    sha256: 1c1222298da2100128f821c57096c69cb6cec0d22ba3b66c2859ae95ae473799
+    patches:
+        - setup.py.patch
```
Comment (jakirkham) [inline recipes/keyring/meta.yaml]: Could we get an explanation of the patch? Looks sensible to me, but a comment here would be nice. Also, maybe open an issue upstream about adding these options.

**Comment by jaraco** [comment]
I'm glad for keyring to be included. I'm not interested in any additional maintenance tasks, though I'm happy to accept a deploy hook for the project to trigger or upload builds as they're released, which for PyPI are released via Travis-CI.

I'm surprised the Windows build doesn't work. I worked primarily on Python 3 on Windows until about a year ago and relied heavily on keyring. Also, the builds for Windows are pure Python, so a Windows build should be straightforward. I know pywin32-ctypes is abandoned, but for the purpose of keyring should be adequate. Keyring alternately supports (or used to support) pywin32, so that may be a suitable expectation for a packaging system.

Hope that helps.

**Comment by ericdill** [comment]
> I'm glad for keyring to be included.

Great!

>  I'm not interested in any additional maintenance tasks

No problem. Thanks for being clear

> though I'm happy to accept a deploy hook for the project to trigger or upload builds as they're released

I'm not that is very straightforward.  I'm specifically thinking of the case when a requirement changes since the meta.yaml (as it stands now) needs to be edited by hand.  Would be great if we could get to a point where such automated builds via deploy hooks were easy.

> I'm surprised the Windows build doesn't work

I guess I need to clarify what the problem actually is.  In building a recipe for keyring, I noticed that there was this pywin32-ctypes requirement for windows and so needed to also build a conda package for that.  One of the community guidelines for conda-forge is to run the test suite, if possible, during the conda build.  The test suite for pywin32-ctypes-v0.0.1 fails on python 3.5 (and maybe all python 3's) with two import errors and I just gave up when I saw that.  It's actually quite simple to patch, but now I'm hitting a different test issue and so am giving up again.

> Keyring alternately supports (or used to support) pywin32, so that may be a suitable expectation for a packaging system.

Thanks for the pointer.  I got this working locally on win64 and py35. Let's see how AppVeyor does with the other build combinations.  Assuming that this works, would you accept a PR that changes the requirement in setup.py from `pywin32-ctypes` to `pywin32`?

**Comment by ericdill** [comment]
Ok looks like this is working on windows with the pywin32 package, yay!

@jakirkham I renamed the patches to be more clear.  I can still add a comment in line if you'd like.

File: recipes/keyring/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    license: Python Software Foundation License or MIT License
+    summary: 'Store and access your passwords safely.'
+
+extra:
+    recipe-maintainers:
+        - ericdill
+        - licode
+        - tacaswell
+
+# See
```
Comment (ocefpaf) [inline recipes/keyring/meta.yaml]: You can remove these comments.

**Comment by ocefpaf** [comment]
LGTM

**Comment by jakirkham** [comment]
> I renamed the patches to be more clear. I can still add a comment in line if you'd like.

Would be nice, but the names are pretty clear and so are the patches for the most part. It's not a big deal just a "nice to have".

**Comment by ocefpaf** [comment]
I was going to ask for the comments, but then I realized that I would have just copy-n-paste the patches names :-)

**Comment by ericdill** [comment]
Thanks all for the help here.  Good to have this in!

**Comment by jakirkham** [comment]
Figured this isn't WIP any more. 😉

**Comment by jaraco** [comment]
> would you accept a PR that changes the requirement in setup.py from pywin32-ctypes to pywin32?

I'd rather not for the Python package as found on PyPI. `pywin32` is a very large package with extension modules, requiring compilation. Furthermore, last I checked, it didn't cooperate very well with `pip` or `easy_install` packaging tools. I'd like to keep the dependency lightweight and extension free if possible.

**Pull request merged**